### ISSUE TITLE
this all looks wrong to me

### ISSUE THREAD

**Issue by gavinking** [adrianmsmith/hibernate-mysql-reconnect-demo#2]
Why would you not just use `hibernate.c3p0.timeout` and/or `hibernate.c3p0.idle_test_period` (or even `hibernate.c3p0.maxConnectionAge`)?

These are well-documented settings of Hibernate, mentioned in both the Hibernate docs and the c3p0 docs. Did you try them?

**Comment by gavinking** [comment]
By the way @adrianmsmith, this blog post:

<https://www.databasesandlife.com/automatic-reconnect-from-hibernate-to-mysql/>

Is the first hit on Google for the phrase "reconnect Hibernate MySQL". If it's incorrect, as I believe, I would appreciate if you would post an update explaining the correct solution to the problem (i.e. setting an appropriate timeout for idle pooled connections).

**Comment by adrianmsmith** [comment]
Hi Gavin, Thanks for your message. I’m on vacation at the moment. I will
look at this next week. That is indeed my blog post, i would be happy to
update it if it’s incorrect.

**Comment by adrianmsmith** [comment]
@gavinking -

There are many reasons why an application might lose a connection to the database. This can happen due to no activity and MySQL dropping the connection after 8 hours, but it can also happen unexpectedly due to temporary network issues, or simply because the database is rebooted. Not even the 8 hours can be relied upon, the database team might change the number of hours to a different value and not inform anyone.

Obviously if the connection drops during the processing of a request there is no hope and the request must be abandoned and an error shown to the user. But if the connection drops temporarily between one request and the next, I see no reason why the second request should not work.

So it seems to me that the `c3p0.testConnectionOnCheckout=true` is still the right one to force that behaviour.

I have looked at the options you mentioned and I don't think any of them solve the above requirement. I think that https://www.mchange.com/projects/c3p0/ are the current official docs. The say:

* `hibernate.c3p0.timeout` (`c3p0.maxIdleTime`) - _"Seconds a Connection can remain pooled but unused before being discarded"_
* `hibernate.c3p0.idle_test_period` (`c3p0.idleConnectionTestPeriod`) - _"c3p0 will test all idle, pooled but unchecked-out connections, every this number of seconds"_
* `hibernate.c3p0.maxConnectionAge` - _"A Connection older than maxConnectionAge will be destroyed and purged from the pool."_

Thanks again for your feedback. Let me know if I've misunderstood that.

**Comment by gavinking** [comment]
OK well that wasn't the scenario described in the blog post, which talked specifically about idle timeouts. Sure, you're right that if 

1. the database is rebooted at a time when most of the connections happen to be idle because there is almost no load in the system, and 
2. you don't what to restart the app itself (which I think is what most people would do),

then always testing the connection before use might be a reasonable thing to do.

Then in that case I believe you can just set the Hibernate property `hibernate.c3p0.testConnectionOnCheckout=true` and you're in business. Is that not right?

**Comment by gavinking** [comment]
(Note that I'm not endorsing this solution in all circumstances, since it requires an extra database round trip for every single transaction.)

**Comment by adrianmsmith** [comment]
Thanks for the correction that I only need the only `hibernate.c3p0.testConnectionOnCheckout=true` property. I have removed all the other properties from the blog post and the code on GitHub.

I don't know why I tried so many properties before, nor why I thought that property needed to go into its own `c3p0.properties` files as opposed to the usual Hibernate place. Maybe things have changed since I originally wrote the post (2008), or maybe I was just confused and got it wrong. Anyway, I've updated it now.

You are right that the post is not very clear that I wish to reconnect no matter what the reason for the database connection drop is. Originally I discovered the problem with the MySQL 8-hour connection timeout, so that triggered the post. I have altered the post to try to clarify that a bit although I fear it's still not very clear.

I agree with that testing on every connection checkout is a performance trade-off. Lower performance in general, but a lower number of requests to the software delivering errors.

I also agree that, these days, an environment such as AWS would restart a piece of software that had failed (e.g. database connection issues). That wasn't the case with the environment I was using back then: once the application went down, it stayed down until someone restarted it. That was a serious problem. Even today, I'm not sure what the right trade-off is between: Simplicity (application crashes and is restarted) vs Resilience (application re-connects and carries on). I think as it's a single config line, I'd go for Resilience here, but absolutely accept others might make other trade-offs. (Also i worry that if the DB restarting causes the application to restart, and if anything with a persistent connection to the application took the same approach then it too must restart, there is a risk of quite a cascade of restarts.)

Let me know if you have any more thoughts.

P.S. I enjoyed your book "Hibernate in Action: Practical Object/Relational Mapping" very much.
